### PR TITLE
[#15] Backends config implementation

### PR DESCRIPTION
**Pull request by DK318**
Problem: there was no config file for backends. All backend stuff should be hardcoded in order to use `coffer`.
Solution: used @MagicRB data types and functions from `Config` module.

### CODE REVIEW

**Comment by MagicRB** [comment]
Hope the API I created isn't too bad :D

File: config.toml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+main_backend = "main"
+
+[[backend]]
+  type = "vault-kv"
+  name = "main"
+  address = "127.0.0.1:8200"
+  mount = "secret"
+  token = "s.XxzTITDlzOsNNzBlYRHURna6"
```
Comment (dcastro) [inline config.toml:8]: The default config file shouldn't have any specific token. I think a placeholder like `"<vault token>"` would do.

File: config.toml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+main_backend = "main"
+
+[[backend]]
```
Comment (dcastro) [inline config.toml:3]: Why the double square brackets `[[` and `]]`?

File: config.toml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+main_backend = "main"
```
Comment (dcastro) [inline config.toml:1]: This file should be added to the `extra-source-files` section in `package.yaml`.

**Review by dcastro — CHANGES_REQUESTED** [review]

File: config.toml (hunk 2 above)
Comment (MagicRB) [reply config.toml:3]: Because it's a list backends, in haskell it creates `[Backend]`

Comment (dcastro) [reply config.toml:3]: Gotcha, thanks!

File: app/Main.hs [hunk 4]
```diff
@@ -11,8 +11,11 @@ import Polysemy
 import Servant.Client (BaseUrl(..), Scheme(Http))
```
Comment (dcastro) [inline app/Main.hs:11]: Some of these imports are redundant.
Please run the pipeline against your PR.

File: config.toml (hunk 3 above)
Comment (dcastro) [reply config.toml]: Why was `config.toml` deleted?

File: app/Main.hs (hunk 4 above)
Comment (dcastro) [inline app/Main.hs:11]: The file `test.toml` should be deleted (or rather, replaced with/renamed to `config.toml`)

File: app/Main.hs [hunk 5]
```diff
@@ -46,104 +44,118 @@ runVaultIO_ action =
       Left err -> do
         die $ show err
 
+readConfig :: IO Config
+readConfig = do
+  text <- TIO.readFile "./config.toml"
+  let config = Toml.decode configCodec text
+  case config of
+    Left err -> TIO.putStrLn (Toml.prettyTomlDecodeErrors err) >> exitFailure
```
Comment (dcastro) [inline app/Main.hs:52]: Please also print a header line like `An error occurred while decoding <config filename>:`

File: app/Main.hs [hunk 6]
```diff
@@ -25,19 +28,14 @@ import Backend
...
-
-mount :: Text
-mount = "secret"
-
-runVaultIO_ :: Sem '[BackendEffect, Error CofferError, Embed IO, Final IO ] a -> IO a
-runVaultIO_ action =
-  runVaultIO url token mount action
+import Config (configCodec, Config (..))
+
+runVaultIO_
```
Comment (dcastro) [inline app/Main.hs:33]: I guess this function has now been generalized to work with _any_ backend (not necessarily vault's backend), so maybe we should rename it to something more generic like `runBackend`.

**Review by dcastro — CHANGES_REQUESTED** [review]

File: config.toml (hunk 1 above)
Comment (DK318) [reply config.toml]: Done

File: config.toml (hunk 3 above)
Comment (DK318) [reply config.toml]: Done

File: app/Main.hs (hunk 4 above)
Comment (DK318) [reply app/Main.hs]: Done

Comment (DK318) [reply app/Main.hs]: Done

File: app/Main.hs (hunk 5 above)
Comment (DK318) [reply app/Main.hs]: Done

File: app/Main.hs (hunk 6 above)
Comment (DK318) [reply app/Main.hs]: Done

File: app/Main.hs [hunk 7]
```diff
@@ -46,104 +41,118 @@ runVaultIO_ action =
       Left err -> do
         die $ show err
 
+readConfig :: IO Config
+readConfig = do
+  text <- TIO.readFile "./config.toml"
+  let config = Toml.decode configCodec text
+  case config of
+    Left err -> TIO.putStrLn ("An error occurred while decoding 'config.toml':" <> Toml.prettyTomlDecodeErrors err) >> exitFailure
```
Comment (dcastro) [inline app/Main.hs:49]: Please print the header msg ("An error occurred ...") and the actual error in separate lines.

If they're printed in the same line, and the actual error msg spans multiple lines, it's gonna look weird.

```suggestion
    Left err -> do
      TIO.putStrLn "An error occurred while decoding 'config.toml':"
      TIO.putStrLn $ Toml.prettyTomlDecodeErrors err
      exitFailure
```

**Review by dcastro — APPROVED** [review]

**Pull request merged**